unittest/buildtool/source_commands_test.py

```diff
@@ -33,12 +35,132 @@
+
+        factory = registry["new_release_branch"]
+        factory.init_argparser(self.parser, defaults)
+
+        options = self.parser.parse_args(["new_release_branch"])
+
+        mock_push_branch = self.patch_method(GitRunner, "push_branch_to_origin")
+
+        command = factory.make_command(options)
+        command()
+
+        base_git_dir = os.path.join(options.input_dir, "new_release_branch")
+        self.assertEqual(os.listdir(base_git_dir), [NORMAL_SERVICE])
+        git_dir = os.path.join(base_git_dir, NORMAL_SERVICE)
+
+        # new 'release' branch HEAD should match 'master' branch HEAD
+        self.assertEqual(
+            GitRunner(options).query_local_repository_commit_id(git_dir),
+            self.repo_commit_map[NORMAL_SERVICE]["master"],
+        )
+
+        # new 'release' branch should have tag
+        self.assertIsNotNone(
+            GitRunner(options).query_commit_at_tag(git_dir, BASE_VERSION_TAG)
+        )
+
+        mock_push_branch.assert_called_once_with(git_dir, "release-0.1.x")
+
+    def test_new_release_branch_with_commits_command(self):
+        """assert new release branch creation:
+        when source branch HEAD is not tagged but a tag exists, successfully branch at latest tag
+        """
+
+        defaults = {
+            "input_dir": self.options.input_dir,
+            "output_dir": self.options.output_dir,
+            "only_repositories": NORMAL_SERVICE,
+            "github_owner": "default",
+            "git_branch": PATCH_BRANCH,  # this branch has commits since last tag
+            "new_branch": "release-0.1.x",
+            "github_repository_root": self.options.github_repository_root,
+        }
+
+        registry = {}
+        buildtool_main.add_standard_parser_args(self.parser, defaults)
+        buildtool.source_commands.register_commands(registry, self.subparsers, defaults)
+
+        factory = registry["new_release_branch"]
+        factory.init_argparser(self.parser, defaults)
+
+        options = self.parser.parse_args(["new_release_branch"])
+
+        mock_push_branch = self.patch_method(GitRunner, "push_branch_to_origin")
+
+        command = factory.make_command(options)
+        command()
+
+        base_git_dir = os.path.join(options.input_dir, "new_release_branch")
+        self.assertEqual(os.listdir(base_git_dir), [NORMAL_SERVICE])
+        git_dir = os.path.join(base_git_dir, NORMAL_SERVICE)
+
+        # new 'release' branch HEAD should match 'master' branch HEAD (which is tagged)
+        self.assertEqual(
+            GitRunner(options).query_local_repository_commit_id(git_dir),
+            self.repo_commit_map[NORMAL_SERVICE]["master"],
+        )
+
+        # new 'release' branch should have tag
+        self.assertIsNotNone(
+            GitRunner(options).query_commit_at_tag(git_dir, BASE_VERSION_TAG)
+        )
+
+        mock_push_branch.assert_called_once_with(git_dir, "release-0.1.x")
+
+    @unittest.expectedFailure
```

Comment:
This decorator asserts that the test case failed for any reason - in our case the failure happens during `command()` because of a branch conflict.
If the test case doesn't fail then the overall test result is a failure per below:
```
Ran 5 tests in 1.853s

FAILED (unexpected successes=1)
```

Reply:
There are a few posts about recording `TestResult` in `tearDown()` but it would need stitching into `BaseGitRepoTestFixture` I think. 

When we run all tests we get told which file had test failures.
When we run a single test file we get told which tests fail providing there are no decorators.
When we use `@unittest.expectedFailure` we don't get told which test failed.

If we change the test fixture in question so that it succeeds, then run the tests and check log output for `unexpected success` we can see last log lines for `test_new_release_branch_exists_command`:
```
D 08:45:49.997 Finished 3719378 with returncode=0 in 0.002 secs
D 08:45:49.998 Running 'git -C "/tmp/TestSourceCommandFixturecsknv1h8/test_new_release_branch_exists_command/input_dir/new_release_branch/gate" rev-parse --abbrev-ref HEAD'...
D 08:45:49.999 Running git as pid 3719379
D 08:45:50.000 Finished 3719379 with returncode=0 in 0.002 secs
D 08:45:50.000 Running 'git -C "/tmp/TestSourceCommandFixturecsknv1h8/test_new_release_branch_exists_command/input_dir/new_release_branch/gate" push origin release-0.1.x.1'...
D 08:45:50.002 Running git as pid 3719380
D 08:45:50.015 Finished 3719380 with returncode=0 in 0.014 secs
I 08:45:50.016 new_release_branch finished gate
D 08:45:50.016 Finished command=new_release_branch
unexpected success
```

Reply:
I'm more concerned with the test failing for the "wrong reason" and not catching it.  Totally fine to leave that for a future improvement though.

Reply:
Ah, I agree it's a bit of a catchall.
I think we can counter that with appropriate test cases for other scenario's. 
Will ponder and submit any extra's as continue wiring up.